src/MagnumPlugins/UfbxImporter/UfbxImporter.cpp

```diff
@@ -0,0 +1,1013 @@
+
+    Vector3 attenuation;
+    LightData::Type lightType;
+
+    if(l->type == UFBX_LIGHT_POINT) {
+        lightType = LightData::Type::Point;
+    } else if(l->type == UFBX_LIGHT_DIRECTIONAL) {
+        lightType = LightData::Type::Directional;
+    } else if(l->type == UFBX_LIGHT_SPOT) {
+        lightType = LightData::Type::Spot;
+    } else {
+        /** @todo area lights */
+        Error() << "Trade::UfbxImporter::light(): light type" << l->type << "is not supported";
+        return {};
+    }
+
+    if (l->decay == UFBX_LIGHT_DECAY_NONE) {
+        attenuation = {1.0f, 0.0f, 0.0f};
+    } else if (l->decay == UFBX_LIGHT_DECAY_LINEAR) {
+        attenuation = {0.0f, 1.0f, 0.0f};
+    } else if (l->decay == UFBX_LIGHT_DECAY_QUADRATIC) {
+        attenuation = {0.0f, 0.0f, 1.0f};
+    } else if (l->decay == UFBX_LIGHT_DECAY_CUBIC) {
+        Warning{} << "Trade::UfbxImporter::light(): cubic attenuation not supported, patching to quadratic";
+        attenuation = {0.0f, 0.0f, 1.0f};
+    } else {
+        Error() << "Trade::UfbxImporter::light(): light type" << l->type << "is not supported";
+    }
+
+
+    if((lightType == LightData::Type::Directional || lightType == LightData::Type::Ambient) && attenuation != Vector3{1.0f, 0.0f, 0.0f}) {
+        Warning{} << "Trade::UfbxImporter::light(): patching attenuation" << attenuation << "to" << Vector3{1.0f, 0.0f, 0.0f} << "for" << lightType;
+        attenuation = {1.0f, 0.0f, 0.0f};
+    }
+
+    return LightData{lightType, color, intensity, attenuation,
+        Deg{(Float)l->inner_angle}, Deg{(Float)l->outer_angle},
+        l};
+}
+
+UnsignedInt UfbxImporter::doMeshCount() const {
+    return (UnsignedInt)_state->meshChunks.size();
+}
+
+Containers::Optional<MeshData> UfbxImporter::doMesh(UnsignedInt id, UnsignedInt level) {
+    if (level != 0) return {};
+
+    MeshChunk chunk = _state->meshChunks[id];
+    ufbx_mesh *mesh = _state->scene->meshes[chunk.meshId];
+    ufbx_mesh_material mat = mesh->materials[chunk.meshMaterialIndex];
+
+    const UnsignedInt indexCount = mat.num_triangles * 3;
+
+    /* Gather all attributes. Position is there always, others are optional */
+    std::size_t attributeCount = 1;
+    std::size_t stride = sizeof(Vector3);
+
+    if(mesh->vertex_normal.exists) {
+        ++attributeCount;
+        stride += sizeof(Vector3);
+    }
+
+    if(mesh->vertex_tangent.exists) {
+        ++attributeCount;
+        stride += sizeof(Vector3);
+    }
+
+    if(mesh->vertex_bitangent.exists) {
+        ++attributeCount;
+        stride += sizeof(Vector3);
+    }
+
+    attributeCount += mesh->uv_sets.count;
+    stride += mesh->uv_sets.count * sizeof(Vector2);
+
+    attributeCount += mesh->color_sets.count;
+    stride += mesh->color_sets.count * sizeof(Color4);
+
+    Containers::Array<UnsignedInt> triangleIndices{mesh->max_face_triangles * 3};
+
+    Containers::Array<char> vertexData{NoInit, stride*indexCount};
```

Comment:
I'm a bit confused about `indexCount` vs `vertexCount`, how do the two relate to each other, and why the vertex data has `indexCount` vertices but the output mesh has `vertexCount` vertices? :)

Reply:
Here the key function is `ufbx_generate_indices()`, it's still lacking documentation but the general idea is as follows:

- Generate 1-N vertex streams for each index (ie. `num_triangles * 3` vertices), here we have one interleaved one
- Call `num_vertices = ufbx_generate_indices(streams, indices, num_indices)` which:
  - Modifies the vertex data pointed by `streams` removing any duplicate vertices
  - Writes the resulting indices to `indices`
  - Returns the number of unique vertices
- Shrink the vertex buffer down to the actually needed `num_vertices`

This step has to be here as FBX files have separate indices for each vertex attribute (kind of like .obj, which is why I said they have similar models) so we can't have a global "index buffer" just from reading a mesh. The indexing also depends on what attributes the user cares about and how they're quantized.

One option is to omit this indexing and leave it up to the users by reporting all duplicated vertices and a consecutive index buffer `{ 0, 1, 2, 3, 4, 5, ... }`, but as other importers already return nicely indexed data I thought I'd include it here, especially as ufbx has the utility function for it.

Reply:
Ah, I see now. Thanks for the clarification. 

What I'm doing in the shitty `ObjImporter` is similar, although I'm not touching any vertex data, operating just on the separate index buffers themselves (as it's always less data to compare, and in case there were any deliberately duplicated vertices with the same attribute data, they get preserved). The output is deduplicated index tuples, which I then use to fill the actual vertex buffer. 

I'm not yet sure which of the two I'd prefer, in any case relevant bits in the `ObjImporter` code would be [here](https://github.com/mosra/magnum/blob/a49602987499379e4a2d155472961d99ddfc75ba/src/MagnumPlugins/ObjImporter/ObjImporter.cpp#L250-L252) (where it's taking a shortcut because it currently supports at most 3 vertex streams) and a [deduplication step here](https://github.com/mosra/magnum/blob/a49602987499379e4a2d155472961d99ddfc75ba/src/MagnumPlugins/ObjImporter/ObjImporter.cpp#L433-L482) (where it again takes a shortcut and assumes positions are always `[0]` in the deduplicated 2D index view, normals `[1]` etc). The strided array views might be a bit overwhelming at first, [this post could hopefully help more than the docs themselves](https://blog.magnum.graphics/backstage/multidimensional-strided-array-views/).

Reply:
Deduplicating by indices would be a lot neater if FBX exporters weren't so cursed. For some reason many applications write normals *always* as non-indexed, ufbx uses a shared contiguous index buffer here so that addressing code stays the same. This means as soon as you include normal data you lose all possibility of indexing as every normal is duplicated (even when smooth normals are enabled).

Reply:
Ugh! Cursed indeed. In that case, the go-to convenience API would be [MeshTools::removeDuplicates()](https://doc.magnum.graphics/magnum/namespaceMagnum_1_1MeshTools.html#ac39584845d0291262e8276e05d53b22e). A `MeshData` goes in, a deduplicated `MeshData` goes out. 

An option to not do this (and return the original duplicated mesh instead) might make sense. 

For OBJ I eventually plan to add support with one index buffer per attribute, which would then add a possibility to just return the input index data unchanged, for the users to have a possibility to bathe in the cursedness in full.

Reply:
Yea I assumed this would be the case, would you prefer to do the deduplication via `MeshTools` rather than ufbx's internal one? That would at least make it possible to create the `StridedArrayView1D` beforehand as they're not resized, though I came up with a way to do it pretty cleanly even using `ufbx_generate_indices()`.

Reply:
Oh also, the deduplication could probably be an importer option at least for now, maybe enabled by default for user convenience?

Reply:
`MeshTools` dependency would be fine in this case, I think it's better to reuse existing functionality than have each plugin implement its own variant of vertex deduplication to avoid that dependency. The `ObjImporter` depends on `MeshTools` for the same reason.

Do that by default, and have an option to disable that if needed, yes.